### ISSUE TITLE
about training strategy

### ISSUE THREAD

**Issue by mati1994** [YantaoShen/kpm_rw_person_reid#1]
I trained the model(with both KPM and GRSW) end to end on Market1501 with options similar to yours with Market-1501 pretrained models for about 10 epochs, but have got a very low mAP(<10%). I noticed that the final BatchNorm layer would not easily converge to a globally optimal value, and also thought that the random walk strategies may be not directly used on the start of strategies(this is just my guess). Would you like to kindly share some about your training strategies, or how many stages have you adopted to retrain your model? Thank you very much.

**Comment by mati1994** [comment]
I finally made it to train the model from pretrained model end-to-end with ~15 epochs and a temporary first stage mAP of 64.0%, while the training is still running and loss has much space to optimize(currently 0.076, and a result with a first stage mAP of 74.9% has loss of around 0.025). There are still much space to optimize, and thank you for your kind work to open source. 
    Besides, I've found several bugs in the code, including torch.randperm() in sampler.py, which causes wrong training sampling, and somewhere for the checkpoint loading. By the way, to the author, have you considered putting pow(*,2) operation before global pooling in p2g embedding module? I think that's more resonable and have adopted it.

**Comment by YantaoShen** [comment]
Hi, Thank you for your interest in our project.

1) I think it is OK to use torch.randperm() in the sampler.py, which randomly shuffles the training data. I have checked the sampling result and it is right. 

2) For checkpoint loading, I am sorry for my lazy coding, the warning is redundant (actually the pre-trained model is loaded successfully). I will update the code later when I have the time. 

3) I tried putting pow(*,2) operation before GAP at the beginning of this project, not as good as putting it after GAP.

**Comment by mati1994** [comment]
hi，thank you for your reply. About the randperm，I think it is possibly due to the version of pytorch that a bug has occurred，and I found that in my environment，when index i is iterated within enumerate(torch.randperm)，i is not python builtin int，which causes a conversion error when it is used as index for defaultdict，and the samples would be wrongly selected. I use pytorch0.4.1 and python3.6.

**Comment by YantaoShen** [comment]
Yes, our code is just tested under python 2.7 and pytorch 0.3.1. Maybe different versions of python and pytorch will give different results.
